pkg/server/container_execsync.go

```diff
@@ -140,40 +131,33 @@ func (c *criContainerdService) execInContainer(ctx context.Context, id string, o
 		}
 	})
 
-	resCh := make(chan execResult, 1)
-	go func() {
-		// Wait will return if context is cancelled.
-		exitCode, err := process.Wait(ctx)
-		resCh <- execResult{
-			exitCode: exitCode,
-			err:      err,
-		}
-		glog.V(2).Infof("Exec process %q exits with exit code %d and error %v", execID, exitCode, err)
-	}()
+	exitCh, err := process.Wait(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait for process %v", err)
+	}
 	if err := process.Start(ctx); err != nil {
 		return nil, fmt.Errorf("failed to start exec %q: %v", execID, err)
 	}
 
-	var timeoutCh <-chan time.Time
-	if opts.timeout == 0 {
-		// Do not set timeout if it's 0.
-		timeoutCh = make(chan time.Time)
-	} else {
-		timeoutCh = time.After(opts.timeout)
+	if opts.timeout != 0 {
+		ctx, cancel = context.WithTimeout(ctx, opts.timeout)
+		defer cancel()
 	}
 	select {
-	case <-timeoutCh:
+	case <-ctx.Done():
+		// If parent context cancel, the ctx done chanel would be closed as well.
 		// Ignore the not found error because the process may exit itself before killing.
-		if err := process.Kill(ctx, unix.SIGKILL); err != nil && !errdefs.IsNotFound(err) {
+		if _, err := process.Delete(ctx, containerd.WithProcessKill); err != nil && !errdefs.IsNotFound(err) {
 			return nil, fmt.Errorf("failed to kill exec %q: %v", execID, err)
 		}
 		// Wait for the process to be killed.
-		<-resCh
+		<-exitCh
```

Comment:
Can we add a log here explain there is a timeout, and the exitCode, error we get here.
I think V(2) should be good enough.

I should have done that before. Thanks!